docs/reference/world-id-state-replication.mdx

```diff
@@ -0,0 +1,174 @@
+
+Full implementation can be found [here](https://github.com/rarimo/worldid-integration-contracts/blob/master/contracts/IdentityManager.sol).
+
+### `Verifier`
+
+<!-- TODO: add Deployments -->
+
+It's an implementation of a verifier contract.
+You can add your business logic to it by modifying (or overriding) two hooks – `_beforeProofSubmit` and `_afterProofSubmit`;
+
+
+#### Interface
+
+```solidity
+contract Verifier is IVerifier, Initializable {
+
+    address public semaphoreVerifier;
+    address public identityManager;
+
+    function verifyProof(
+        uint256 root_,
+        uint256 signalHash_,
+        uint256 nullifierHash_,
+        uint256 externalNullifierHash_,
+        uint256[8] calldata proof_
+    ) public virtual override {
+        // ...
+
+        _beforeProofValidation();
+
+        ISemaphoreVerifier(semaphoreVerifier).verifyProof(
+            proof_,
+            [root_, nullifierHash_, signalHash_, externalNullifierHash_]
+        );
+
+        _afterProofValidation();
+    }
+
+    function _beforeProofValidation() internal virtual {}
+
+    function _afterProofValidation() internal virtual {}
+}
+
+```
+
+- `semaphoreVerifier` — address of the `SemaphoreVerifier` contract that verifies the proof;
+- `identityManager` — address of the `IdentityManager` contract;
+- `verifyProof(...)` —  the main function that verifies the user's proof.
+It calls the `IdentityManager` to verify whether the root is actual and `SemaphoreVerifier` to verify the proof itself;
+- `_beforeProofValidation()` — hook that should be executed before the `verifyProof(...)`;
+- `_afterProofValidation()` — hook that gets executed after the `verifyProof(...)` call;
+
+Full implementation can be found [here](https://github.com/rarimo/worldid-integration-contracts/blob/master/contracts/Verifier.sol).
+
+### `SemaphoreVerifier`
+
+<!-- TODO: add Deployments -->
+
+This contract is responsible for verifying the ZKP.
+It has some necessary math functions, constants, and the primary `verifyProof(...)` function.
+It allows proofs to be verified in uncompressed or compressed form.
+Usually, uncompressed form is used, see [this](https://2π.com/23/bn254-compression) for further explanation.
+
+#### Interface
+
+```solidity
+contract SemaphoreVerifier is ISemaphoreVerifier {
+
+    /* ... */
+
+    function verifyCompressedProof(
+        uint256[4] calldata compressedProof,
+        uint256[4] calldata input
+    ) public view {
+        // ...
+    }
+
+    function verifyProof(uint256[8] calldata proof, uint256[4] calldata input) public view {
+        // ...
+    }
+}
+```
+
+- `verifyCompressedProof(...)` —  verifies a compressed proof;
+- `verifyProof(...)` — verifies an uncompressed proof;
+
+Full implementation can be found [here](https://github.com/rarimo/worldid-integration-contracts/blob/master/contracts/vendor/worldcoin/world-id-contracts/SemaphoreVerifier.sol).
```

Comment:
```suggestion
For the full implementation, see [SemaphoreVerifier.sol](https://github.com/rarimo/worldid-integration-contracts/blob/master/contracts/vendor/worldcoin/world-id-contracts/SemaphoreVerifier.sol).
```
I want to introduce this style of references into our docs:
https://developers.google.com/style/cross-references#references-to-other-documents

Please update all the external links in this doc accordingly